// @bone/bone-mobile-ui/lib/ActionSheet/lib/ActionSheet.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Modal = require('../../Modal');

var _Modal2 = _interopRequireDefault(_Modal);

var _Content = require('./Content');

var _Content2 = _interopRequireDefault(_Content);

var _constants = require('../../util/constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  open: function open() {
    var _this = this;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        title = _ref.title,
        titleStyle = _ref.titleStyle,
        _ref$options = _ref.options,
        options = _ref$options === undefined ? [] : _ref$options,
        _ref$onPress = _ref.onPress,
        onPress = _ref$onPress === undefined ? _constants.noop : _ref$onPress,
        cancelText = _ref.cancelText,
        cancelStyle = _ref.cancelStyle,
        _ref$onCancelPress = _ref.onCancelPress,
        onCancelPress = _ref$onCancelPress === undefined ? _constants.noop : _ref$onCancelPress,
        _ref$closeable = _ref.closeable,
        closeable = _ref$closeable === undefined ? true : _ref$closeable;

    var modal = _Modal2.default.open({
      type: 'popup',
      duration: 0,
      closeable: closeable,
      onClose: _constants.noop,
      transparent: false,
      children: _react2.default.createElement(_Content2.default, {
        title: title,
        titleStyle: titleStyle,
        options: options,
        onPress: function () {
          var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(index) {
            return _regenerator2.default.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.prev = 0;
                    _context.next = 3;
                    return onPress(index);

                  case 3:
                    modal.close();
                    _context.next = 8;
                    break;

                  case 6:
                    _context.prev = 6;
                    _context.t0 = _context['catch'](0);

                  case 8:
                  case 'end':
                    return _context.stop();
                }
              }
            }, _callee, _this, [[0, 6]]);
          }));

          return function (_x2) {
            return _ref2.apply(this, arguments);
          };
        }(),
        cancelText: cancelText,
        cancelStyle: cancelStyle,
        onCancelPress: (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
          return _regenerator2.default.wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  _context2.prev = 0;
                  _context2.next = 3;
                  return onCancelPress();

                case 3:
                  modal.close();
                  _context2.next = 8;
                  break;

                case 6:
                  _context2.prev = 6;
                  _context2.t0 = _context2['catch'](0);

                case 8:
                case 'end':
                  return _context2.stop();
              }
            }
          }, _callee2, _this, [[0, 6]]);
        }))
      })
    });
    return { close: function close() {
        return modal.close();
      } };
  }
};
