// @bone/bone-mobile-ui/lib/List/lib/List.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var List = function (_Component) {
  (0, _inherits3.default)(List, _Component);

  function List() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, List);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = List.__proto__ || (0, _getPrototypeOf2.default)(List)).call.apply(_ref, [this].concat(args))), _this), _this.swipeStore = [], _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(List, [{
    key: 'renderChildren',
    // store to cache swipe opened list item

    // consider the children maybe like bellow:
    // [childA, [childB, childC, [childD, childE]]]
    // actually, they're still the same children layer
    value: function renderChildren(children) {
      var _this2 = this;

      var index = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      var lastIndex = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

      if (Array.isArray(children)) {
        return children.map(function (child, index) {
          return _this2.renderChildren(child, index, children.length - 1);
        });
      } else if (_react2.default.isValidElement(children)) {
        return _react2.default.cloneElement(children, {
          noBorder: index === lastIndex,
          swipeStore: this.swipeStore,
          key: children.key || 'key-bone-mobile-ui-list-item-' + index
        });
      } else {
        return null;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          style = _props.style,
          children = _props.children;

      return _react2.default.createElement(
        _reactNative.View,
        { style: style },
        this.renderChildren(children)
      );
    }
  }]);
  return List;
}(_react.Component);

List.Item = null;
List.propTypes = {
  style: _reactNative.View.propTypes.style,
  children: _propTypes2.default.node
};
exports.default = List;
