// @bone/bone-mobile-ui/lib/Header/lib/Header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Header = function (_Component) {
  (0, _inherits3.default)(Header, _Component);

  function Header() {
    (0, _classCallCheck3.default)(this, Header);
    return (0, _possibleConstructorReturn3.default)(this, (Header.__proto__ || (0, _getPrototypeOf2.default)(Header)).apply(this, arguments));
  }

  (0, _createClass3.default)(Header, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          style = _props.style,
          title = _props.title,
          titleStyle = _props.titleStyle;

      return _react2.default.createElement(
        _reactNative.View,
        { style: [styles.body, style] },
        _react2.default.createElement(
          _reactNative.Text,
          { style: [styles.title, titleStyle] },
          title
        )
      );
    }
  }]);
  return Header;
}(_react.Component);

Header.propTypes = {
  style: _reactNative.View.propTypes.style,
  title: _propTypes2.default.string,
  titleStyle: _reactNative.Text.propTypes.style
};
exports.default = Header;


var styles = _reactNative.StyleSheet.create({
  body: {
    minHeight: 36,
    paddingTop: 9,
    paddingRight: 16,
    paddingBottom: 9,
    paddingLeft: 16
  },
  title: {
    lineHeight: 18,
    fontSize: 12,
    color: '#999'
  }
});
