// @bone/intl/lib/core/intl.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.intlShape = undefined;
exports.default = createIntl;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createIntl(locale, messages) {
  (0, _invariant2.default)(typeof locale === 'string', 'Expected locale to be a string');
  (0, _invariant2.default)((0, _util.isPlainObject)(messages), 'Expected messages to be a plain object');

  var currentLocale = locale;
  var currentMessages = messages;
  var currentListeners = [];

  var getMessage = function getMessage() {
    return currentMessages[currentLocale] || {};
  };

  var subscribe = function subscribe(listener) {
    (0, _invariant2.default)(typeof listener === 'function', 'Expected listener to be a function');

    var isSubscribed = true;
    currentListeners.push(listener);

    return function unsubscribe() {
      if (!isSubscribed) {
        return;
      }

      isSubscribed = false;
      currentListeners.splice(currentListeners.indexOf(listener), 1);
    };
  };

  var dispatch = function dispatch(action) {
    (0, _invariant2.default)((0, _util.isPlainObject)(action), 'Expected action to be a plain object');

    if ('locale' in action) {
      (0, _invariant2.default)(typeof action.locale === 'string', 'Expected action.locale to be a string');
      currentLocale = action.locale;
    }

    if ('messages' in action) {
      (0, _invariant2.default)((0, _util.isPlainObject)(action.messages), 'Expected action.messages to be a plain object');
      currentMessages = action.messages;
    }

    currentListeners.forEach(function (listener) {
      return listener();
    });
    return action;
  };

  return {
    getMessage: getMessage,
    subscribe: subscribe,
    dispatch: dispatch
  };
}

var intlShape = exports.intlShape = _propTypes2.default.shape({
  getMessage: _propTypes2.default.func.isRequired,
  subscribe: _propTypes2.default.func.isRequired,
  dispatch: _propTypes2.default.func.isRequired
}).isRequired;
