// @bone/bone-mobile-ui/lib/util/createRootSibling/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRootSibling;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _inject = require('./inject');

var _inject2 = _interopRequireDefault(_inject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var id = 0;
var offStreamElement = function offStreamElement(el) {
  return _react2.default.cloneElement(el, { style: [el.props.style, styles.offStream] });
};

function createRootSibling(el, callback) {
  var currentId = id++;
  var update = function update(el, callback) {
    return _inject2.default.emit('siblingupdate', currentId, offStreamElement(el), callback);
  };
  var destroy = function destroy(callback) {
    return _inject2.default.emit('siblingupdate', currentId, null, callback);
  };
  update(el, callback);
  return { update: update, destroy: destroy };
}

var styles = _reactNative.StyleSheet.create({
  offStream: {
    position: 'absolute'
  }
});
