// @bone/bone-mobile-ui/lib/Gesture/lib/joystick.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createJoystickHandler;

var _constants = require('../../util/constants');

function createJoystickHandler(emitter) {
  function onTouchStart(el, evt) {
    var _el$props$onJoystickS = el.props.onJoystickStart,
        onJoystickStart = _el$props$onJoystickS === undefined ? _constants.noop : _el$props$onJoystickS;

    onJoystickStart(calculateJoystickState(el, evt));

    // add joystick listeners
    emitter.on('touchmove', onJoystick);
    emitter.on('touchend', onJoystickEnd);
  }

  function onJoystick(el, evt) {
    var _el$props$onJoystick = el.props.onJoystick,
        onJoystick = _el$props$onJoystick === undefined ? _constants.noop : _el$props$onJoystick;

    onJoystick(calculateJoystickState(el, evt));
  }

  function onJoystickEnd(el, evt) {
    // remove joystick listeners
    removeEventListeners();

    var _el$props$onJoystickE = el.props.onJoystickEnd,
        onJoystickEnd = _el$props$onJoystickE === undefined ? _constants.noop : _el$props$onJoystickE;

    onJoystickEnd(calculateJoystickState(el, evt));
  }

  function removeEventListeners() {
    emitter.removeListener('touchmove', onJoystick);
    emitter.removeListener('touchend', onJoystickEnd);
  }

  function calculateJoystickState(el, evt) {
    var _el$joystickData = el.joystickData,
        startX = _el$joystickData.startX,
        startY = _el$joystickData.startY,
        width = _el$joystickData.width,
        height = _el$joystickData.height;
    var _evt$nativeEvent = evt.nativeEvent,
        locationX = _evt$nativeEvent.locationX,
        locationY = _evt$nativeEvent.locationY,
        pageX = _evt$nativeEvent.pageX,
        pageY = _evt$nativeEvent.pageY;

    var radian = Math.PI - Math.atan2(pageX - startX - width / 2, pageY - startY - height / 2);
    var angle = radian / Math.PI * 180;
    return {
      locationX: locationX, locationY: locationY, pageX: pageX, pageY: pageY,
      angle: angle, radian: radian
    };
  }

  return { onTouchStart: onTouchStart };
}
