// @bone/bone-mobile-ui/lib/Picker/lib/Gradient.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _GradientBase2 = require('./GradientBase');

var _GradientBase3 = _interopRequireDefault(_GradientBase2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Gradient = function (_GradientBase) {
  (0, _inherits3.default)(Gradient, _GradientBase);

  function Gradient() {
    (0, _classCallCheck3.default)(this, Gradient);
    return (0, _possibleConstructorReturn3.default)(this, (Gradient.__proto__ || (0, _getPrototypeOf2.default)(Gradient)).apply(this, arguments));
  }

  (0, _createClass3.default)(Gradient, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          height = _props.height,
          reverse = _props.reverse;

      return _react2.default.createElement(
        _reactNative.View,
        { style: styles.body },
        _react2.default.createElement('div', {
          style: {
            height: height,
            background: reverse ? 'linear-gradient(to bottom, rgba(255, 255, 255, 0.6), rgba(255, 255, 255, 0.95))' : 'linear-gradient(to bottom, rgba(255, 255, 255, 0.95), rgba(255, 255, 255, 0.6))'
          }
        })
      );
    }
  }]);
  return Gradient;
}(_GradientBase3.default);

exports.default = Gradient;


var styles = _reactNative.StyleSheet.create({
  body: {
    flex: 1
  }
});
