// @bone/bone-mobile-ui/lib/Toast/lib/Info.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Info = function (_Component) {
  (0, _inherits3.default)(Info, _Component);

  function Info() {
    (0, _classCallCheck3.default)(this, Info);
    return (0, _possibleConstructorReturn3.default)(this, (Info.__proto__ || (0, _getPrototypeOf2.default)(Info)).apply(this, arguments));
  }

  (0, _createClass3.default)(Info, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          width = _props.width,
          title = _props.title,
          titleStyle = _props.titleStyle;

      return _react2.default.createElement(
        _reactNative.View,
        { style: [styles.body, { width: width }] },
        _react2.default.createElement(
          _reactNative.Text,
          { style: [styles.title, titleStyle] },
          title
        )
      );
    }
  }]);
  return Info;
}(_react.Component);

Info.propTypes = {
  width: _propTypes2.default.number,
  title: _propTypes2.default.string,
  titleStyle: _reactNative.Text.propTypes.style
};
Info.defaultProps = {
  width: _constants.TOAST_MIN_WIDTH,
  title: ''
};
exports.default = Info;


var styles = _reactNative.StyleSheet.create({
  body: {
    minWidth: _constants.TOAST_MIN_WIDTH,
    maxWidth: _constants.TOAST_MAX_WIDTH,
    paddingTop: 34,
    paddingRight: 20,
    paddingBottom: 34,
    paddingLeft: 20,
    borderRadius: 6,
    backgroundColor: 'rgba(35, 35, 35, 0.7)'
  },
  title: {
    lineHeight: 22,
    fontSize: 16,
    color: '#fff',
    textAlign: 'center'
  }
});
