// @bone/bone-mobile-ui/lib/Toast/lib/Toast.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Modal = require('../../Modal');

var _Modal2 = _interopRequireDefault(_Modal);

var _Info = require('./Info');

var _Info2 = _interopRequireDefault(_Info);

var _Loading = require('./Loading');

var _Loading2 = _interopRequireDefault(_Loading);

var _constants = require('../../util/constants');

var _constants2 = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  info: function info() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$duration = _ref.duration,
        duration = _ref$duration === undefined ? _constants2.DEFAULT_TOAST_DURATION : _ref$duration,
        _ref$onClose = _ref.onClose,
        onClose = _ref$onClose === undefined ? _constants.noop : _ref$onClose,
        _ref$width = _ref.width,
        width = _ref$width === undefined ? _constants2.TOAST_MIN_WIDTH : _ref$width,
        _ref$title = _ref.title,
        title = _ref$title === undefined ? '' : _ref$title,
        titleStyle = _ref.titleStyle;

    var modal = _Modal2.default.open({
      type: 'toast',
      duration: duration,
      closeable: false,
      onClose: onClose,
      transparent: true,
      children: _react2.default.createElement(_Info2.default, { width: width, title: title, titleStyle: titleStyle })
    });
    return { close: function close() {
        return modal.close();
      } };
  },
  loading: function loading() {
    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref2$duration = _ref2.duration,
        duration = _ref2$duration === undefined ? _constants2.DEFAULT_TOAST_DURATION : _ref2$duration,
        _ref2$onClose = _ref2.onClose,
        onClose = _ref2$onClose === undefined ? _constants.noop : _ref2$onClose,
        _ref2$width = _ref2.width,
        width = _ref2$width === undefined ? _constants2.LOADING_MIN_WIDTH : _ref2$width,
        _ref2$title = _ref2.title,
        title = _ref2$title === undefined ? '' : _ref2$title,
        titleStyle = _ref2.titleStyle;

    var modal = _Modal2.default.open({
      type: 'toast',
      duration: duration,
      closeable: false,
      onClose: onClose,
      transparent: true,
      children: _react2.default.createElement(_Loading2.default, { width: width, title: title, titleStyle: titleStyle })
    });
    return { close: function close() {
        return modal.close();
      } };
  }
};
