// @bone/bone-mobile-ui/lib/Alert/lib/Alert.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Modal = require('../../Modal');

var _Modal2 = _interopRequireDefault(_Modal);

var _Content = require('./Content');

var _Content2 = _interopRequireDefault(_Content);

var _constants = require('../../util/constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  open: function open() {
    var _this = this;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        title = _ref.title,
        titleStyle = _ref.titleStyle,
        message = _ref.message,
        messageStyle = _ref.messageStyle,
        _ref$actions = _ref.actions,
        actions = _ref$actions === undefined ? [] : _ref$actions;

    var modal = _Modal2.default.open({
      type: 'alert',
      duration: 0,
      closeable: false,
      onClose: _constants.noop,
      transparent: false,
      children: _react2.default.createElement(_Content2.default, {
        title: title,
        titleStyle: titleStyle,
        message: message,
        messageStyle: messageStyle,
        actions: actions.map(function (action) {
          return (0, _extends3.default)({}, action, {
            onPress: function () {
              var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
                var _action$onPress, onPress;

                return _regenerator2.default.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _action$onPress = action.onPress, onPress = _action$onPress === undefined ? _constants.noop : _action$onPress;
                        _context.prev = 1;
                        _context.next = 4;
                        return onPress();

                      case 4:
                        modal.close();
                        _context.next = 9;
                        break;

                      case 7:
                        _context.prev = 7;
                        _context.t0 = _context['catch'](1);

                      case 9:
                      case 'end':
                        return _context.stop();
                    }
                  }
                }, _callee, _this, [[1, 7]]);
              }));

              function onPress() {
                return _ref2.apply(this, arguments);
              }

              return onPress;
            }()
          });
        })
      })
    });
    return { close: function close() {
        return modal.close();
      } };
  }
};
