// @bone/bone-mobile-ui/lib/Modal/lib/DefaultContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _Container2 = require('./Container');

var _Container3 = _interopRequireDefault(_Container2);

var _TouchableOpacity = require('../../TouchableOpacity');

var _TouchableOpacity2 = _interopRequireDefault(_TouchableOpacity);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DefaultContainer = function (_Container) {
  (0, _inherits3.default)(DefaultContainer, _Container);

  function DefaultContainer() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, DefaultContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = DefaultContainer.__proto__ || (0, _getPrototypeOf2.default)(DefaultContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      visible: false
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(DefaultContainer, [{
    key: 'open',


    // redefine open
    value: function open() {
      this.setState({ visible: true });
    }

    // redefine close

  }, {
    key: 'close',
    value: function close() {
      var onAfterClose = this.props.onAfterClose;

      this.setState({ visible: false });
      onAfterClose();
    }

    // redefine render

  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          onOverlayPress = _props.onOverlayPress,
          transparent = _props.transparent,
          children = _props.children;
      var visible = this.state.visible;

      return visible ? _react2.default.createElement(
        _reactNative.View,
        { style: styles.body },
        _react2.default.createElement(_TouchableOpacity2.default, {
          style: [styles.background, { backgroundColor: transparent ? 'transparent' : _themes2.default['color.overlay.modal'] }],
          activeOpacity: 1,
          onPress: onOverlayPress
        }),
        children
      ) : null;
    }
  }]);
  return DefaultContainer;
}(_Container3.default);

exports.default = DefaultContainer;


var styles = _reactNative.StyleSheet.create({
  body: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    backgroundColor: 'transparent',
    overflow: 'hidden'
  },
  background: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  }
});
