// @bone/bone-mobile-ui/lib/Modal/lib/PopupContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _Container2 = require('./Container');

var _Container3 = _interopRequireDefault(_Container2);

var _TouchableOpacity = require('../../TouchableOpacity');

var _TouchableOpacity2 = _interopRequireDefault(_TouchableOpacity);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _Dimensions$get = _reactNative.Dimensions.get('window'),
    height = _Dimensions$get.height;

var PopupContainer = function (_Container) {
  (0, _inherits3.default)(PopupContainer, _Container);

  function PopupContainer() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, PopupContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = PopupContainer.__proto__ || (0, _getPrototypeOf2.default)(PopupContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      visible: false,
      opacity: new _reactNative.Animated.Value(0),
      translateY: new _reactNative.Animated.Value(0)
    }, _this.contentHeight = 0, _this.onAfterClose = function () {
      var onAfterClose = _this.props.onAfterClose;

      _this.setState({ visible: false });
      onAfterClose();
    }, _this.onContentLayout = function (e) {
      try {
        _this.contentHeight = e.nativeEvent.layout.height || 0;
      } catch (err) {
        _this.contentHeight = 0;
      }
      _this.forceUpdate();
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(PopupContainer, [{
    key: 'stopAnimate',
    value: function stopAnimate() {
      var _state = this.state,
          opacity = _state.opacity,
          translateY = _state.translateY;

      opacity.stopAnimation();
      translateY.stopAnimation();
    }

    // redefine open

  }, {
    key: 'open',
    value: function open() {
      var _state2 = this.state,
          opacity = _state2.opacity,
          translateY = _state2.translateY;

      this.stopAnimate();
      this.setState({ visible: true });
      _reactNative.Animated.parallel([_reactNative.Animated.timing(opacity, { duration: _constants.POPUP_ANIMATION_DURATION, toValue: 1, easing: _reactNative.Easing.linear }), _reactNative.Animated.timing(translateY, { duration: _constants.POPUP_ANIMATION_DURATION, toValue: 1, easing: _reactNative.Easing.in })]).start();
    }

    // redefine close

  }, {
    key: 'close',
    value: function close() {
      var _state3 = this.state,
          opacity = _state3.opacity,
          translateY = _state3.translateY;

      this.stopAnimate();
      _reactNative.Animated.parallel([_reactNative.Animated.timing(opacity, { duration: _constants.POPUP_ANIMATION_DURATION, toValue: 0, easing: _reactNative.Easing.linear }), _reactNative.Animated.timing(translateY, { duration: _constants.POPUP_ANIMATION_DURATION, toValue: 0, easing: _reactNative.Easing.in })]).start(this.onAfterClose);
    }
  }, {
    key: 'render',


    // redefine render
    value: function render() {
      var _props = this.props,
          onOverlayPress = _props.onOverlayPress,
          transparent = _props.transparent,
          children = _props.children;
      var _state4 = this.state,
          visible = _state4.visible,
          opacity = _state4.opacity,
          translateY = _state4.translateY;

      return visible ? _react2.default.createElement(
        _reactNative.View,
        { style: styles.body },
        _react2.default.createElement(
          _reactNative.Animated.View,
          {
            style: [styles.background, {
              backgroundColor: transparent ? 'transparent' : _themes2.default['color.overlay.modal'],
              opacity: opacity
            }]
          },
          _react2.default.createElement(_TouchableOpacity2.default, { style: styles.backgroundTouch, activeOpacity: 1, onPress: onOverlayPress })
        ),
        _react2.default.createElement(
          _reactNative.Animated.View,
          {
            style: [styles.container, {
              transform: [{
                translateY: translateY.interpolate({
                  inputRange: [0, 1],
                  outputRange: [0, -this.contentHeight]
                })
              }]
            }],
            onLayout: this.onContentLayout
          },
          children
        )
      ) : null;
    }
  }]);
  return PopupContainer;
}(_Container3.default);

exports.default = PopupContainer;


var styles = _reactNative.StyleSheet.create({
  body: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    backgroundColor: 'transparent',
    overflow: 'hidden'
  },
  background: {
    flex: 1
  },
  backgroundTouch: {
    flex: 1
  },
  container: {
    position: 'absolute',
    top: height,
    right: 0,
    left: 0
  }
});
