// @bone/bone-mobile-ui/lib/Modal/lib/ToastContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _Container2 = require('./Container');

var _Container3 = _interopRequireDefault(_Container2);

var _TouchableOpacity = require('../../TouchableOpacity');

var _TouchableOpacity2 = _interopRequireDefault(_TouchableOpacity);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ToastContainer = function (_Container) {
  (0, _inherits3.default)(ToastContainer, _Container);

  function ToastContainer() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ToastContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ToastContainer.__proto__ || (0, _getPrototypeOf2.default)(ToastContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      visible: false,
      opacity: new _reactNative.Animated.Value(0)
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ToastContainer, [{
    key: 'stopAnimate',
    value: function stopAnimate() {
      var opacity = this.state.opacity;

      opacity.stopAnimation();
    }

    // redefine open

  }, {
    key: 'open',
    value: function open() {
      var opacity = this.state.opacity;

      this.stopAnimate();
      this.setState({ visible: true });
      _reactNative.Animated.timing(opacity, {
        duration: _constants.TOAST_ANIMATION_DURATION,
        toValue: 1,
        easing: _reactNative.Easing.linear
      }).start();
    }

    // redefine close

  }, {
    key: 'close',
    value: function close() {
      var onAfterClose = this.props.onAfterClose;
      var opacity = this.state.opacity;

      this.stopAnimate();
      this.setState({ visible: false });
      opacity.setValue(0);
      onAfterClose();
    }

    // redefine render

  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          onOverlayPress = _props.onOverlayPress,
          transparent = _props.transparent,
          children = _props.children;
      var _state = this.state,
          visible = _state.visible,
          opacity = _state.opacity;

      return visible ? _react2.default.createElement(
        _reactNative.View,
        { style: styles.body },
        _react2.default.createElement(_TouchableOpacity2.default, {
          style: [styles.background, { backgroundColor: transparent ? 'transparent' : _themes2.default['color.overlay.modal'] }],
          activeOpacity: 1,
          onPress: onOverlayPress
        }),
        _react2.default.createElement(
          _reactNative.Animated.View,
          { style: { opacity: opacity } },
          children
        )
      ) : null;
    }
  }]);
  return ToastContainer;
}(_Container3.default);

exports.default = ToastContainer;


var styles = _reactNative.StyleSheet.create({
  body: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    backgroundColor: 'transparent',
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'center'
  },
  background: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  }
});
