// @bone/bone-mobile-ui/lib/Radio/lib/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Radio = function (_Component) {
  (0, _inherits3.default)(Radio, _Component);

  function Radio() {
    (0, _classCallCheck3.default)(this, Radio);
    return (0, _possibleConstructorReturn3.default)(this, (Radio.__proto__ || (0, _getPrototypeOf2.default)(Radio)).apply(this, arguments));
  }

  (0, _createClass3.default)(Radio, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          checked = _props.checked,
          disabled = _props.disabled,
          color = _props.color;

      return _react2.default.createElement(
        _reactNative.View,
        { style: [styles.body, { opacity: checked ? 1 : 0 }] },
        _react2.default.createElement(_reactNative.View, { style: [styles.check, { borderColor: disabled ? _themes2.default['radio.color.disable'] : color }] })
      );
    }
  }]);
  return Radio;
}(_react.Component);

Radio.propTypes = {
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  color: _propTypes2.default.string
};
Radio.defaultProps = {
  checked: false,
  disabled: false,
  color: '#1fc88b'
};
exports.default = Radio;


var styles = _reactNative.StyleSheet.create({
  body: {
    width: 24,
    height: 24,
    alignItems: 'center',
    justifyContent: 'center'
  },
  check: {
    width: 19,
    height: 10,
    borderBottomWidth: 3,
    borderLeftWidth: 3,
    transform: [{ rotate: '-45deg' }, { translateX: 1 }, { translateY: -1 }]
  }
});
