// @bone/bone-mobile-ui/lib/Checkbox/lib/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _themes = require('../../style/themes');

var _themes2 = _interopRequireDefault(_themes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Checkbox = function (_Component) {
  (0, _inherits3.default)(Checkbox, _Component);

  function Checkbox() {
    (0, _classCallCheck3.default)(this, Checkbox);
    return (0, _possibleConstructorReturn3.default)(this, (Checkbox.__proto__ || (0, _getPrototypeOf2.default)(Checkbox)).apply(this, arguments));
  }

  (0, _createClass3.default)(Checkbox, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          checked = _props.checked,
          disabled = _props.disabled,
          color = _props.color;

      return _react2.default.createElement(
        _reactNative.View,
        {
          style: [styles.body, {
            borderColor: disabled ? _themes2.default['checkbox.color.disable'] : color,
            backgroundColor: checked ? disabled ? _themes2.default['checkbox.color.disable'] : color : _themes2.default['color.fill.general']
          }]
        },
        _react2.default.createElement(_reactNative.View, { style: styles.check })
      );
    }
  }]);
  return Checkbox;
}(_react.Component);

Checkbox.propTypes = {
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  color: _propTypes2.default.string
};
Checkbox.defaultProps = {
  checked: false,
  disabled: false,
  color: '#1fc88b'
};
exports.default = Checkbox;


var styles = _reactNative.StyleSheet.create({
  body: {
    width: 24,
    height: 24,
    borderRadius: 12,
    borderWidth: 2,
    alignItems: 'center',
    justifyContent: 'center'
  },
  check: {
    width: 13,
    height: 7,
    borderBottomWidth: 2,
    borderLeftWidth: 2,
    borderColor: _themes2.default['color.fill.general'],
    transform: [{ rotate: '-45deg' }, { translateX: 1 }, { translateY: -1 }]
  }
});
