// @bone/bone-mobile-ui/lib/Flex/lib/Flex.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Flex = function (_Component) {
  (0, _inherits3.default)(Flex, _Component);

  function Flex() {
    (0, _classCallCheck3.default)(this, Flex);
    return (0, _possibleConstructorReturn3.default)(this, (Flex.__proto__ || (0, _getPrototypeOf2.default)(Flex)).apply(this, arguments));
  }

  (0, _createClass3.default)(Flex, [{
    key: 'translateFlexPropValue',
    value: function translateFlexPropValue(value) {
      switch (value) {
        case 'start':
          return 'flex-start';
        case 'end':
          return 'flex-end';
        case 'between':
          return 'space-between';
        case 'around':
          return 'space-around';
        default:
          return value;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          direction = _props.direction,
          align = _props.align,
          justify = _props.justify,
          wrap = _props.wrap,
          style = _props.style,
          children = _props.children;

      var flexStyle = {
        flexDirection: direction,
        alignItems: this.translateFlexPropValue(align),
        justifyContent: this.translateFlexPropValue(justify),
        flexWrap: wrap
      };
      return _react2.default.createElement(
        _reactNative.View,
        { style: [flexStyle, style] },
        children
      );
    }
  }]);
  return Flex;
}(_react.Component);

Flex.Item = null;
Flex.propTypes = {
  direction: _propTypes2.default.oneOf(['row', 'column']),
  align: _propTypes2.default.oneOf(['start', 'end', 'center', 'stretch']),
  justify: _propTypes2.default.oneOf(['start', 'end', 'center', 'between', 'around']),
  wrap: _propTypes2.default.oneOf(['wrap', 'nowrap']),
  style: _reactNative.View.propTypes.style,
  children: _propTypes2.default.node
};
Flex.defaultProps = {
  direction: 'column',
  align: 'stretch',
  justify: 'start',
  wrap: 'nowrap'
};
exports.default = Flex;
